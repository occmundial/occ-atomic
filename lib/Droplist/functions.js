import unorm from 'unorm';

export const compareText = (text = '', term = '') => {
	const accents = /[\u0300-\u036f]/g;
	const textLC = unorm.nfd(text.toLowerCase().replace(accents, ""));
	const termLC = unorm.nfd(term.toLowerCase().replace(accents, ""));
	const index = textLC.indexOf(termLC);
	return index;
};

export const separateText = (itemText = "", index, term) => {
	const text = [itemText.substr(0, index), itemText.substr(index, term.length), itemText.substr(index+term.length, itemText.length)];
	return text;
};
