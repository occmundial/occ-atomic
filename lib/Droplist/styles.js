import colors from '../subatomic/colors';
import spacing from '../subatomic/spacing';
const { gutter, tiny, small, base, xTiny } = spacing;
const { primLighter, grey50, grey200, bgWhite,textLink} = colors;
export default {
	block: {
		background: bgWhite,
		border: `1px solid ${grey200}`,
		borderRadius: xTiny,
		overflow: 'hidden'
	},
	group: {
		padding: [small, small, xTiny]
	},
	item: {
		display: 'flex',
		justifyContent: 'space-between',
		padding: [tiny, small],
		transition: '0.1s all',
		cursor: 'pointer',
		position: 'relative',
		'&:hover': {
			background: grey50
		}
	},
	right: {
		float: 'right',
		transition: '0.1s all'
	},
	onFocus: {
		background: primLighter,
		'&:hover': {
			background: primLighter
		}
	},
	icon: {
		marginRight: tiny,
		position: 'absolute',
		top: gutter
	},
	iconText: {
		display: 'inline-block',
		marginLeft: base
	},
	extraText:{
		fontWeight: 600,
		color: textLink,
		marginLeft: tiny
	}
};
