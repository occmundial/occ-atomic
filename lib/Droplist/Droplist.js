import React from 'react';
import PropTypes from 'prop-types';
import autobind from 'class-autobind';
import isEqual from 'lodash/isEqual';
import cloneDeep from 'lodash/cloneDeep';
import Text from '../Text';
import Icon from '../Icon';
import Colors from '../subatomic/colors';
import { compareText, separateText } from './functions';
import iconSizes from '../subatomic/iconSizes';

const arrowDown = 40;
const arrowUp = 38;
const enter = 13;
const { inkLighter } = Colors;
const { small: iconSmall } = iconSizes;
/**
 * The Droplist component displays a list and filters it with the prop 'term'.
 * The value of 'term' es highlighted in every item that matches.
 * The array of objects needed to display the items needs a text and an id, and can contain a text that floats on the right. If you're going to group your items by category, you first need to set an array of groups, and each group must contain an array with the items.
 */
class Droplist extends React.Component {
	constructor(props) {
		super(props);
		autobind(this);
		this.state = {
			currentItem: -1,
			currentGroup: -1,
			items: props.items
		};
	}

	componentDidMount() {
		const { items, filter, term } = this.props;
		window.addEventListener('keydown', this.onKeyDown);
		if (filter) this.filterItems(items, term);
	}

	componentWillReceiveProps(props) {
		if (
			props.term !== this.props.term ||
			!isEqual(props.items, this.props.items)
		) {
			this.setState({ currentItem: -1 });
			if (props.filter) this.filterItems(props.items, props.term);
			else this.setState({ items: props.items });
		}
	}

	componentWillUnmount() {
		window.removeEventListener('keydown', this.onKeyDown);
	}

	onKeyDown(e) {
		const { items } = this.state;
		const { isOnFocus, groups } = this.props;
		if (isOnFocus) {
			if (e.which === arrowUp || e.which === arrowDown) {
				e.preventDefault();
				if (groups) {
					if (e.which === arrowDown) {
						this.moveGroupDown();
					} else {
						this.moveGroupUp();
					}
				} else {
					if (e.which === arrowDown) {
						this.moveDown(items);
					} else {
						this.moveUp(items);
					}
				}
			}
			if (e.which === enter) {
				this.onEnter();
			}
		}
	}

	onEnter() {
		const { currentItem, currentGroup, items } = this.state;
		const { onEnter, groups, groupItemsKey } = this.props;
		if (currentItem > -1) {
			if (groups) {
				const selectedGroup = items.find((group, i) => i === currentGroup);
				const selectedItem = selectedGroup[groupItemsKey].find(
					(item, i) => i === currentItem
				);
				this.setState({ currentGroup: -1, currentItem: -1 });
				if (onEnter) onEnter(selectedItem);
			} else {
				const selectedItem = items.find((item, i) => i === currentItem);
				this.setState({ currentItem: -1 });
				if (onEnter) onEnter(selectedItem);
			}
		}
	}

	moveGroupDown() {
		const { currentItem, currentGroup, items } = this.state;
		const { groupItemsKey } = this.props;
		if (currentGroup === -1) {
			this.moveToNextGroup();
		} else {
			const groupItems = items[currentGroup][groupItemsKey];
			if (currentItem === groupItems.length - 1) {
				this.moveToNextGroup();
			} else {
				this.moveDown(groupItems);
			}
		}
	}

	moveGroupUp() {
		const { currentItem, currentGroup, items } = this.state;
		const { groupItemsKey } = this.props;
		if (currentGroup === -1) {
			this.moveToPreviousGroup();
		} else {
			const groupItems = items[currentGroup][groupItemsKey];
			if (currentItem === 0) {
				this.moveToPreviousGroup();
			} else {
				this.moveUp(groupItems);
			}
		}
	}

	moveToNextGroup() {
		const { currentGroup, items } = this.state;
		const { groupItemsKey } = this.props;
		if (currentGroup === items.length - 1) {
			this.setState({ currentGroup: -1, currentItem: -1 });
		} else {
			this.setState({ currentGroup: currentGroup + 1 }, () => {
				const { currentGroup } = this.state;
				if (
					items[currentGroup][groupItemsKey] &&
					items[currentGroup][groupItemsKey].length
				) {
					this.setState({ currentItem: 0 });
				} else {
					this.moveToNextGroup();
				}
			});
		}
	}

	moveToPreviousGroup() {
		const { currentGroup, items } = this.state;
		const { groupItemsKey } = this.props;
		if (currentGroup === -1) {
			this.setState({ currentGroup: items.length - 1 }, () => {
				const { currentGroup } = this.state;
				if (
					items[currentGroup][groupItemsKey] &&
					items[currentGroup][groupItemsKey].length
				) {
					this.setState({
						currentItem: items[currentGroup][groupItemsKey].length - 1
					});
				} else {
					this.moveToPreviousGroup();
				}
			});
		} else if (currentGroup === 0) {
			this.setState({ currentGroup: -1, currentItem: -1 });
		} else {
			this.setState({ currentGroup: currentGroup - 1 }, () => {
				const { currentGroup } = this.state;
				if (
					items[currentGroup][groupItemsKey] &&
					items[currentGroup][groupItemsKey].length
				) {
					this.setState({
						currentItem: items[currentGroup][groupItemsKey].length - 1
					});
				} else {
					this.moveToPreviousGroup();
				}
			});
		}
	}

	moveDown(items) {
		const { currentItem } = this.state;
		if (currentItem === items.length - 1) {
			this.setState({ currentItem: -1 });
		} else {
			this.setState({ currentItem: currentItem + 1 });
		}
	}

	moveUp() {
		const { currentItem, items } = this.state;
		if (currentItem === -1) {
			this.setState({ currentItem: items.length - 1 });
		} else {
			this.setState({ currentItem: currentItem - 1 });
		}
	}

	onClick(item, e) {
		const { onClick } = this.props;
		e.stopPropagation();
		this.setState({ currentGroup: -1, currentItem: -1 });
		if (onClick) onClick(item);
	}

	onMouseDown(item, e) {
		const { onMouseDown } = this.props;
		e.stopPropagation();
		this.setState({ currentGroup: -1, currentItem: -1 });
		if (onMouseDown) onMouseDown(item);
	}

	onMouseUp(item, e) {
		const { onMouseUp } = this.props;
		e.stopPropagation();
		this.setState({ currentGroup: -1, currentItem: -1 });
		if (onMouseUp) onMouseUp(item);
	}

	filterItems(items, term) {
		const { itemTextKey, groupItemsKey, groups } = this.props;
		if (groups) {
			let newItems = cloneDeep(items);
			newItems = items.map(group => {
				group[groupItemsKey] = group[groupItemsKey].filter(
					item => compareText(item[itemTextKey], term) >= 0
				);
				return group;
			});
			newItems = newItems.filter(group => group[groupItemsKey].length > 0);
			this.setState({ items: newItems });
		} else {
			const newItems = items.filter(
				item => compareText(item[itemTextKey], term) >= 0
			);
			this.setState({ items: newItems });
		}
	}

	renderList(items, selectedGroup) {
		const { currentItem } = this.state;
		const {
			classes,
			itemIdKey,
			itemTextKey,
			itemTextRightKey,
			term
		} = this.props;
		const itemsDOM = items.map((item, i) => {
			let index = compareText(item[itemTextKey], term);
			if (index >= 0) {
				let text = separateText(item[itemTextKey], index, term);
				return (
					<div
						key={item[itemIdKey]}
						onClick={e => this.onClick(item, e)}
						onMouseDown={e => this.onMouseDown(item, e)}
						onMouseUp={e => this.onMouseUp(item, e)}
						className={`${classes.item}${
							selectedGroup && currentItem === i ? ` ${classes.onFocus}` : ''
						}`}
					>
						<div>
							{item.iconName && (
								<Icon
									iconName={item.iconName}
									width={iconSmall}
									height={iconSmall}
									display="inline-block"
									className={classes.icon}
									colors={[inkLighter]}
								/>
							)}
							<Text className={item.iconName ? classes.iconText : ''}>
								{text[0].length ? text[0] : ''}
								<Text tag="b" strong>
									{text[1].length ? text[1] : ''}
								</Text>
								{text[2].length ? text[2] : ''}
								{item.extraText && (
									<span className={classes.extraText}>{item.extraText}</span>
								)}
							</Text>
						</div>
						{item[itemTextRightKey] && (
							<span className={classes.right}>
								<Text tag="span" low>
									{item[itemTextRightKey]}
								</Text>
							</span>
						)}
					</div>
				);
			} else {
				return (
					<div
						key={item[itemIdKey]}
						onClick={e => this.onClick(item, e)}
						onMouseDown={e => this.onMouseDown(item, e)}
						onMouseUp={e => this.onMouseUp(item, e)}
						className={`${classes.item}${
							selectedGroup && currentItem === i ? ` ${classes.onFocus}` : ''
						}`}
					>
						<div>
							{item.iconName && (
								<Icon
									iconName={item.iconName}
									width={14}
									height={14}
									display="inline-block"
									className={classes.dsfsdf}
									colors={[inkLighter]}
								/>
							)}
							<Text>
								{item[itemTextKey]}
								{item.extraText && (
									<span className={classes.extraText}>{item.extraText}</span>
								)}
							</Text>
						</div>
						{item[itemTextRightKey] && (
							<span className={classes.right}>
								<Text tag="span" low>
									{item[itemTextRightKey]}
								</Text>
							</span>
						)}
					</div>
				);
			}
		});
		return itemsDOM;
	}

	render() {
		const { items, currentGroup } = this.state;
		const {
			classes,
			className,
			style,
			id,
			groups,
			groupNameKey,
			groupIdKey,
			groupItemsKey
		} = this.props;
		return (
			<div
				className={`${classes.block}${className ? ` ${className}` : ''}`}
				id={id}
				style={style}
			>
				{groups
					? items.map((group, index) => {
							return (
								<div key={group[groupIdKey]}>
									<Text small mid className={classes.group}>
										{group[groupNameKey].toUpperCase()}
									</Text>
									{this.renderList(
										group[groupItemsKey],
										currentGroup === index
									)}
								</div>
							);
					  })
					: this.renderList(items, true)}
			</div>
		);
	}
}

Droplist.defaultProps = {
	groups: false,
	groupNameKey: 'text',
	groupIdKey: 'id',
	groupItemsKey: 'items',
	itemTextKey: 'text',
	itemTextRightKey: 'textRight',
	itemIdKey: 'id',
	term: '',
	filter: true
};

Droplist.propTypes = {
	classes: PropTypes.object.isRequired,
	/** Array of objects with all the items to display. */
	items: PropTypes.array,
	/** Term to filter the items of the Droplist. */
	term: PropTypes.string,
	/** Key to find in the object, to use as the text. */
	itemTextKey: PropTypes.string,
	/** Key to find in the object, to use as the text that floats on the right. */
	itemTextRightKey: PropTypes.string,
	/** Key to find in the object, to use as the id. */
	itemIdKey: PropTypes.string,
	/** Use this if the items are grouped in different categories. */
	groups: PropTypes.bool,
	/** Key to find in the object to use as the name of the group. */
	groupNameKey: PropTypes.string,
	/** Key to find in the object to use as the id of the group. */
	groupIdKey: PropTypes.string,
	/** Key to find in the object where you can find the array of items. */
	groupItemsKey: PropTypes.string,
	/** Function to call when an item is clicked. */
	onClick: PropTypes.func,
	/** Function to call when mouse is pressed down on an item. */
	onMouseDown: PropTypes.func,
	/** Function to call when mouse is pressed up on an item. */
	onMouseUp: PropTypes.func,
	/** Function to call when the Enter key is pressed down. */
	onEnter: PropTypes.func,
	/** Filters given items if they match with the term. */
	filter: PropTypes.bool,
	/** With this parameter the Droplist will have navigation with up and down keys. */
	isOnFocus: PropTypes.bool,
	id: PropTypes.string,
	className: PropTypes.string,
	style: PropTypes.object
};

export default Droplist;
