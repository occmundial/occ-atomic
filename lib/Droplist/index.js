import Droplist from './Droplist';

export default Droplist;
