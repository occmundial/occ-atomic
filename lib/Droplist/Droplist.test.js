import React from 'react';
import { mount } from 'enzyme';
import { JssProvider, SheetsRegistry } from 'react-jss';

import Droplist from './Droplist';

const items = [
    { text: 'Administrador', textRight: '(222)', id:1 },
    { text: 'Administrador Sr', textRight: '(134)', id:2 },
    { text: 'Administrador Jr', textRight: '(52)', id:3 },
    { text: 'Administrar', textRight: '(6)', id:4 },
];

const items2 = [
    { text: 'Países', id:'1',
        items: [
            { text: 'México', textRight: '(3405)', key: 'MX', type: 'CO', id:'1.1' }
        ]
    },
    { text: 'Estados', id:'2',
        items: [
            { text: 'Querétaro', textRight: '(222)', key: 'QRO', type: 'ST', id:'2.1' },
            { text: 'Estado de México', textRight: '(225)', key: 'EMX', type: 'ST', id:'2.2' }
        ]
    },
    { text: 'Ciudades', id:'3',
        items: [
            { text: 'Santiago de Querétaro', textRight: '(102)', key: 'QRO2', type: 'CY', id:'3.1' }
        ]
    }
];

let sheets, classes;

describe("Droplist", () => {

    beforeEach(() => sheets = new SheetsRegistry());
    afterEach(() => sheets = null);

    it('matches the snapshot', () => {
        const wrapper = mount(
            <Droplist items={items} />
        );
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('returns the right number of items', () => {
        const wrapper = mount(
            <JssProvider registry={sheets}>
                <Droplist items={items} />
            </JssProvider>
        );
        classes = sheets.registry[1].classes;
        expect(wrapper.find(`.${classes.item}`).length).toBe(4);
    });

    it('returns the right number of groups and its items', () => {
        const wrapper = mount(
            <JssProvider registry={sheets}>
                <Droplist items={items2} groups />
            </JssProvider>
        );
        classes = sheets.registry[1].classes;
        const block = `.${classes.block}>div`;
        const item = `.${classes.item}`;
        expect(wrapper.find(block).length).toBe(3);
        expect(wrapper.find(block).at(0).find(item).length).toBe(1);
        expect(wrapper.find(block).at(1).find(item).length).toBe(2);
        expect(wrapper.find(block).at(2).find(item).length).toBe(1);
    });

    it('renders the item correctly', () => {
        const wrapper = mount(
            <JssProvider registry={sheets}>
                <Droplist items={items} itemIdKey="id" />
            </JssProvider>
        );
        classes = sheets.registry[1].classes;
        expect(wrapper.find(`.${classes.item}`).at(0).text()).toBe('Administrador(222)');
    });

    it('renders the group and sub-item correctly', () => {
        const wrapper = mount(
            <JssProvider registry={sheets}>
                <Droplist items={items2} groups groupIdKey="id" itemIdKey="id" />
            </JssProvider>
        );
        classes = sheets.registry[1].classes;
        expect(wrapper.find(`.${classes.block}>div`).at(0).text()).toBe('PAÍSESMéxico(3405)');
    });

    it('filters the items', () => {
        const wrapper = mount(
            <JssProvider registry={sheets}>
                <Droplist items={items} itemIdKey="id" />
            </JssProvider>
        );
        classes = sheets.registry[1].classes;
        const item = `.${classes.item}`;
        expect(wrapper.find(item).length).toBe(4);
        wrapper.setProps({ children: <Droplist items={items} itemIdKey="id" term="Administrador" /> });
        wrapper.instance().forceUpdate();
        wrapper.update();
        expect(wrapper.find(item).length).toBe(3);
    });

    it('highlights the matching part of the items', () => {
        const wrapper = mount(
            <JssProvider registry={sheets}>
                <Droplist items={items} itemIdKey="id" />
            </JssProvider>
        );
        classes = sheets.registry[1].classes;
        const item = `.${classes.item}`;
        expect(wrapper.find(item).at(0).find('b').text()).toBe('');
        wrapper.setProps({ children: <Droplist items={items} itemIdKey="id" term="Administrador" /> });
        wrapper.instance().forceUpdate();
        wrapper.update();
        expect(wrapper.find(item).at(0).find('b').text()).toBe('Administrador');
    });

    it('calls the onClick, onMouseDown and onMouseUp functions', () => {
        const onClick = jest.fn();
        const onMouseDown = jest.fn();
        const onMouseUp = jest.fn();
        const wrapper = mount(
            <JssProvider registry={sheets}>
                <Droplist items={items} itemIdKey="id" onClick={onClick} onMouseDown={onMouseDown} onMouseUp={onMouseUp} />
            </JssProvider>
        );
        classes = sheets.registry[1].classes;
        const item = `.${classes.item}`;
        wrapper.find(item).at(0).simulate('click', {stopPropagation: jest.fn()});
        expect(onClick.mock.calls.length).toBe(1);
        wrapper.find(item).at(0).simulate('mousedown', {stopPropagation: jest.fn()});
        expect(onMouseDown.mock.calls.length).toBe(1);
        wrapper.find(item).at(0).simulate('mouseup', {stopPropagation: jest.fn()});
        expect(onMouseUp.mock.calls.length).toBe(1);
    });

});
