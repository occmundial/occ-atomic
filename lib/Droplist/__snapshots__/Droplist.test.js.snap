// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Droplist matches the snapshot 1`] = `ShallowWrapper {}`;

exports[`Droplist styles matches the snapshot 1`] = `
Object {
  "block": Object {
    "background": "#fff",
    "border": "1px solid #D3D4DC",
    "borderRadius": "4px",
    "boxShadow": "0 9px 12px -5px rgba(8,13,56,0.12), 0 10px 32px 0 rgba(9,61,173,0.12)",
    "overflow": "hidden",
    "padding": Array [
      "8px",
    ],
  },
  "disabled": Object {
    "pointerEvents": "none",
  },
  "extraText": Object {
    "marginLeft": "4px",
  },
  "group": Object {
    "padding": Array [
      "8px",
      "16px",
      0,
    ],
  },
  "highlighted": Object {
    "background": "hsl(221 91.2% 35.7% / 0.05)",
  },
  "icon": Object {
    "marginRight": "8px",
  },
  "iconText": Object {
    "display": "inline-block",
  },
  "item": Object {
    "&:active, &:focus": Object {
      "& > $right": Object {
        "color": "#080D39",
      },
      "background": "hsl(221 91.2% 35.7% / 0.1)",
    },
    "&:hover": Object {
      "background": "hsl(221 91.2% 35.7% / 0.05)",
    },
    "alignItems": "center",
    "borderRadius": "4px",
    "display": "flex",
    "justifyContent": "space-between",
    "padding": Array [
      "12px",
      "16px",
    ],
    "position": "relative",
    "transition": "0.1s all",
  },
  "itemContainer": Object {
    "alignItems": "center",
    "display": "flex",
  },
  "itemPointer": Object {
    "cursor": "pointer",
  },
  "onFocus": Object {
    "&:hover": Object {
      "background": "hsl(221 91.2% 35.7% / 0.1)",
    },
    "background": "hsl(221 91.2% 35.7% / 0.1)",
  },
  "right": Object {
    "float": "right",
    "marginLeft": "16px",
    "transition": "0.1s all",
  },
}
`;

exports[`DroplistJSS matches the snapshot 1`] = `ShallowWrapper {}`;
